usr/lib/common/lock_btree.c

```diff
@@ -379,5 +375,28 @@ void bt_destroy(struct btree *t, void (*delete_func)(void *))
     t->free_list = NULL;
     t->free_nodes = 0;
 
-    pthread_rwlock_unlock(&btree_rwlock);
+    pthread_mutex_unlock(&t->mutex);
+    pthread_mutex_destroy(&t->mutex);
+}
+
+/* bt_init
+ *
+ * Initialize a btree.
+ */
+void bt_init(struct btree *t)
+{
+    pthread_mutexattr_t attr;
+
+    t->free_list = NULL;
+    t->top = NULL;
+    t->size = 0;
+    t->free_nodes = 0;
+
+    /*
+     * Need a recursive mutex, because btree callback functions may call
+     * btree functions again
+     */
+    pthread_mutexattr_init(&attr);
+    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
```

Comment:
Using a lock per btree is certainly the right way to go. The problem with recursive locking is a) kernel has to keep track of a lot of things so its usually slow b) pthreads doesnt know recursive rwlocks so you have to make it a mutex which blocks read ops. And id say once an application has set up everything (sessions, created keys ..), most of th the ops on btrees would be r/o. So if there is a chance we could get rid of btree's recursive calls id consider going with rw locks.

Reply:
The callbacks for freeing a node are not critical, none of these callbacks use btree functions currently. 

The critical callbacks are those for bt_for_each_node(). There are several different callbacks in the code, and some of them do in fact use btree functions inside the callback. 

If they would only do read-type of locks, then a rw lock could still be OK (since rw locks allow multiple readers, even for the same thread), however, I am not sure if the callbacks are only doing read operations. I fear they also do write operations (e.g. bt_node_free(), bt_put_node_value()).

So as of today, I don't see a chance to change this to non-recursive rw locks.

Reply:
im okay with leaving it as is for now. first make it work. we can think about perf improvements later on.